File: 02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {

    FILE *fp = fopen("02.txt", "r");
    if (!fp) exit(EXIT_FAILURE);

    char op[16];
    int x, pos = 0, aim_depth = 0, real_depth = 0;

    do
    {
        fscanf(fp, "%s %d", op, &x);

        if(strcmp(op, "forward") == 0)
        {
            pos += x;
            real_depth += aim_depth * x;
        }

        else if(strcmp(op, "up") == 0)
        {
            aim_depth -= x;
        }

        else if(strcmp(op, "down") == 0)
        {
            aim_depth += x;
        }

    } while (!feof(fp));

    fclose(fp);
    printf("%d\n%d\n", aim_depth * pos, real_depth * pos);
}

File: 02.py
def operations(path="02.txt"):
    with open(path) as lines:
        for op, x in map(str.split, lines):
            yield op, int(x)


def solve():
    real_depth = aim_depth = pos = 0
    # forward is not aim/ depth related
    direction = {
        "up"  : -1,
        "down":  1,
    }

    for op, x in operations():
        if op == "forward":
            pos += x
            real_depth += aim_depth * x
        else:
            aim_depth += x * direction[op]

    return aim_depth * pos, real_depth * pos


p1, p2 = solve()
print(p1)
print(p2)
